fairseq/training.py

```diff
@@ -0,0 +1,491 @@
+            )
+        self.out_plot_se = self.output_path("loss_and_accuracy.svg")
+        self.out_plot_lr = self.output_path("learning_rate.svg")
+
+        # Requirements:
+        self.gpu_rqmt = gpu_rqmt
+        self.rqmt = {
+            "gpu": gpu_rqmt,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+        if self.gpu_rqmt > 1:
+            self.rqmt["cpu"] *= self.gpu_rqmt
+            self.rqmt["mem"] *= self.gpu_rqmt
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+        yield Task("plot", mini_task=True)
+
+    @classmethod
+    def create_fairseq_hydra_config(
+        cls, fairseq_hydra_config, max_epoch, save_interval, **kwargs
+    ):
+        res = copy.deepcopy(fairseq_hydra_config)
+        config_dict = {}
+        post_config_dict = {
+            "optimization": {"max_epoch": max_epoch},
+            "checkpoint": {"save_interval": save_interval},
+        }
+        res.update(FairseqHydraConfig(config_dict, post_config_dict))
+        return res
+
+    def create_files(self):
+        self.fairseq_hydra_config.write(self.out_fairseq_hydra_yaml.get_path())
+        util.create_executable("fairseq.sh", self._get_run_cmd())
+
+    def run(self):
+        if self.use_cache_manager:
+            manifest_path = self.fairseq_hydra_config.config_dict["task"][
+                "data"
+            ].get_path()
+            if self.zipped_audio_dir is None:
+                for name in ["train.tsv", "valid.tsv"]:
+                    with open(os.path.join(manifest_path, name), "r") as manifest_file:
+                        manifest_lines = manifest_file.read().splitlines()
+                    audio_path = manifest_lines[0]
+                    bundle_lines = map(
+                        lambda line: audio_path + "/" + line.split("\t")[0],
+                        manifest_lines[1:],
+                    )
+                    with open(f"{name}.bundle", "w") as bundle_file:
+                        bundle_file.write("\n".join(bundle_lines))
+                    try:
+                        cached_audio_fn = (
+                            sp.check_output(["cf", f"{name}.bundle"])
+                            .strip()
+                            .decode("utf8")
+                        )
+                    except sp.CalledProcessError:
+                        print(f"Cache manager: Error occurred for files in {name}")
+                        raise
+
+                    with open(cached_audio_fn) as local_bundle:
+                        bundle_lines = list(local_bundle.readlines())
+                        manifest_lines[0] = os.path.commonpath(bundle_lines)
+                        manifest_lines[1:] = map(
+                            lambda line: os.path.relpath(line, manifest_lines[0]),
+                            manifest_lines[1:],
+                        )
```

Comment:
@vieting this is my vision how to replace the specific cache manager call (for the bundle file) with the more general sisyphus file_caching call (for each file individually).
(pls. do not just use this diff, it is written in a very small window without actual testing 🙃 )
```suggestion
                    cached_bundle_lines = [gs.file_caching(l) for l in bundle_lines]
                    manifest_lines[0] = os.path.commonpath(cached_bundle_lines )
                    manifest_lines[1:] = map(
                        lambda line: os.path.relpath(line, manifest_lines[0]),
                        manifest_lines[1:],
                    )
```
Note that in some versions of the file_caching function flying around, no actual caching is done, but only `'cf <str> '` is returned.. this would not work here, as fairseq does not interpret these strings. instead the function should directly use the cache manager (either via `sp.check_call(['cf', str])` or by importing the cachemanager python module and then directly using its api)

Reply:
That would be more general, yes. If @Judyxujj could test this, that would be nice. But then, we should also generalize this for a cached zip file below (currently line 290).

Reply:
From what I understand this will not work properly because currently we use this `cf <>` syntax as string in file caching, instead of performing the actual caching. This was very specific for the RASR configs that support this, or jobs that used "shell" execution (which is kind of deprecated now). I am not sure how we can properly fix this design issue in our recipes and Sisyphus itself...

Reply:
Unfortunately not at all, because then we would write the `/var/tmp` path in the rasr config files in the `create_files` step, which would break nearly all setups...

Reply:
that is true.. we would have to merge all "create_config" and "run" steps.. :(

ok, what are other solutions? 
* Keep the i6-specific option in this job and everybody else (i.e. AppTek) will just ignore it?
* An i6-specific wrapper that should be passed to `fairseq_python_exe` that correctly resolves `cf <>` entries in the yaml file?